### PR TITLE
improved and fix bug of getting next batch in mnist

### PR DESCRIPTION
**Pull request by LaurenLuoYun**
The original codes do not consider the situation that when the batch_size > num_examples (even though it is not a common event, if happens, no new data will be input at all).

Actually, a fully epoch means that all examples have been applied once, however, the original codes do not show it. If the batch_size sets just a little bit larger than half of the example number, nearly half of the examples cannot be used in one epoch.

Also, I set the shuffle to be a input variable.

### CODE REVIEW

File: tensorflow/contrib/learn/python/learn/datasets/mnist.py [hunk 1]
```diff
@@ -168,22 +168,40 @@ def next_batch(self, batch_size, fake_data=False):
       return [fake_image for _ in xrange(batch_size)], [
           fake_label for _ in xrange(batch_size)
       ]
+
+    """ Get next batch """ 
     start = self._index_in_epoch
-    self._index_in_epoch += batch_size
-    if self._index_in_epoch > self._num_examples:
+    # Shuffle for the first epoch
+    if self._epochs_completed ==0 and start == 0 and shuffle:
```
Comment (rmlarsen) [inline tensorflow/contrib/learn/python/learn/datasets/mnist.py]: Add space after ==

File: tensorflow/contrib/learn/python/learn/datasets/mnist.py [hunk 2]
```diff
@@ -168,22 +168,40 @@ def next_batch(self, batch_size, fake_data=False):
...
+      perm0 = numpy.arange(self._num_examples)
+      numpy.random.shuffle(perm0)
+      self._images = self.images[perm0]
+      self._labels = self.labels[perm0]
+    # Go to the next epoch
+    if start+batch_size > self._num_examples:
```
Comment (rmlarsen) [inline tensorflow/contrib/learn/python/learn/datasets/mnist.py]: add space around +

File: tensorflow/contrib/learn/python/learn/datasets/mnist.py (hunk 1 above)
Comment (rmlarsen) [inline tensorflow/contrib/learn/python/learn/datasets/mnist.py]: This changes the semantics by shuffling before the first epoch. Is that intentional? Is it necessary?

File: tensorflow/contrib/learn/python/learn/datasets/mnist.py [hunk 3]
```diff
@@ -168,22 +168,40 @@ def next_batch(self, batch_size, fake_data=False):
...
+        numpy.random.shuffle(perm)
+        self._images = self.images[perm]
+        self._labels = self.labels[perm]
       # Start next epoch
       start = 0
-      self._index_in_epoch = batch_size
-      assert batch_size <= self._num_examples
-    end = self._index_in_epoch
-    return self._images[start:end], self._labels[start:end]
+      self._index_in_epoch = batch_size-rest_num_examples
```
Comment (rmlarsen) [inline tensorflow/contrib/learn/python/learn/datasets/mnist.py]: space around '-'

File: tensorflow/contrib/learn/python/learn/datasets/mnist.py [hunk 4]
```diff
@@ -168,22 +168,40 @@ def next_batch(self, batch_size, fake_data=False):
       return [fake_image for _ in xrange(batch_size)], [
           fake_label for _ in xrange(batch_size)
       ]
+
+    """ Get next batch """
```
Comment (rmlarsen) [inline tensorflow/contrib/learn/python/learn/datasets/mnist.py]: Get rid of leading and training space. Add period at end.

**Review by rmlarsen — CHANGES_REQUESTED** [review]

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by LaurenLuoYun** [comment]
@rmlarsen I've already normalized the format, thank you so much and pls check the new pr(43e7a40) again.

File: tensorflow/contrib/learn/python/learn/datasets/mnist.py (hunk 4 above)
Comment (LaurenLuoYun) [inline tensorflow/contrib/learn/python/learn/datasets/mnist.py]: OK , fixed.

File: tensorflow/contrib/learn/python/learn/datasets/mnist.py (hunk 1 above)
Comment (LaurenLuoYun) [inline tensorflow/contrib/learn/python/learn/datasets/mnist.py]: OK , fixed.

Comment (LaurenLuoYun) [inline tensorflow/contrib/learn/python/learn/datasets/mnist.py]: I think it necessary. If not necessary, the shuffle variable can be set to False.

File: tensorflow/contrib/learn/python/learn/datasets/mnist.py (hunk 2 above)
Comment (LaurenLuoYun) [inline tensorflow/contrib/learn/python/learn/datasets/mnist.py]: OK , fixed.

File: tensorflow/contrib/learn/python/learn/datasets/mnist.py (hunk 3 above)
Comment (LaurenLuoYun) [inline tensorflow/contrib/learn/python/learn/datasets/mnist.py]: OK , fixed.

**Comment by LaurenLuoYun** [comment]
@tensorflow-jenkins Follow by the agsignee's instruction, I've change some format of the code. Please test again . Thank you.

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by rmlarsen** [comment]
Thanks for the update. I'll merge this once the  tests pass.

**Comment by LaurenLuoYun** [comment]
@rmlarsen the jenkins test shows that Windows Cmake Tests failure, but I do not fully understand the reason from the output console.

**Review by rmlarsen — APPROVED** [review]

**Comment by rmlarsen** [comment]
The cmake error is unrelated. Thanks for the contribution. I'll merge it now.
